database/email.go

```diff
@@ -0,0 +1,200 @@
+		FailedAttemptsToSend int                `bson:"failed_attempts_to_send"`
+		Failed               bool               `bson:"failed"`
+	}
+)
+
+// EmailCreate creates an email message in the DB which is waiting to be sent.
+func (db *DB) EmailCreate(ctx context.Context, m EmailMessage) error {
+	fields, err := bson.Marshal(m)
+	if err != nil {
+		return err
+	}
+	_, err = db.staticEmails.InsertOne(ctx, fields)
+	if err != nil {
+		return errors.AddContext(err, "failed to Insert")
+	}
+	return nil
+}
+
+// EmailLockAndFetch locks up to batchSize records with the given lockId and
+// returns up to batchSize locked entries. Some of the returned entries might
+// not have been locked during the current execution.
+func (db *DB) EmailLockAndFetch(ctx context.Context, lockID string, batchSize int64) ([]EmailMessage, error) {
+	// Find out how many entries are already locked by this id. Maybe we don't
+	// need to lock any additional ones.
+	filter := bson.M{
+		"locked_by": lockID,
+		"failed":    bson.M{"$ne": true},
+		"sent_at":   nil,
+	}
+	count, err := db.staticEmails.CountDocuments(ctx, filter)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to count locked email messages")
+	}
+	// Lock some more entries in order to fill the batch.
+	if count < batchSize {
+		// As MongoDB doesn't have an "update up to N entries" operation, what
+		// we do here is fetch the ids of the desired number of entries to
+		// lock and then lock them by ids.
+		ids, err := db.fetchUnlockedMessageIDs(ctx, batchSize-count)
+		if err != nil {
+			return nil, errors.AddContext(err, "failed to fetch message ids to lock")
+		}
+		err = db.lockMessages(ctx, lockID, ids)
+		if err != nil {
+			return nil, errors.AddContext(err, "failed to lock messages")
+		}
+	}
+	// Fetch up to batchSize messages already locked with lockID.
+	opts := options.Find()
+	opts.SetLimit(batchSize)
+	c, err := db.staticEmails.Find(ctx, filter, opts)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to fetch emails")
+	}
+	defer func() {
+		if errDef := c.Close(ctx); errDef != nil {
+			db.staticLogger.Traceln("Error on closing DB cursor.", errDef)
+		}
+	}()
+	var msgs []EmailMessage
+	for c.Next(ctx) {
+		var m EmailMessage
+		if err = c.Decode(&m); err != nil {
+			return nil, errors.AddContext(err, "failed to parse value from DB")
+		}
+		msgs = append(msgs, m)
+	}
+	return msgs, nil
+}
+
+// fetchUnlockedMessageIDs is a helper method that fetches the ids of up to num
+// unlocked email messages waiting to be sent.
+func (db *DB) fetchUnlockedMessageIDs(ctx context.Context, num int64) ([]primitive.ObjectID, error) {
+	filter := bson.M{
+		"locked_by": "",
+		"failed":    bson.M{"$ne": true},
+		"sent_at":   nil,
+	}
+	opts := options.Find()
+	opts.SetLimit(num)
+	c, err := db.staticEmails.Find(ctx, filter, opts)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to fetch ids")
+	}
+	defer func() {
+		if errDef := c.Close(ctx); errDef != nil {
```

Comment:
same question on combining errors. Otherwise a comment explaining why not

Reply:
I can add a comment.

Reply:
Combined.